@tclune FYI, I hit an issue as I was testing my latest changes.

```
5 errors found in build log:
     115
     116      275 |            & values, size(values), MPI_INTEGER, &
     117          |             1
     118    ......
     119      298 |            & values_, size(values), MPI_LOGICAL, &
     120          |             2
  >> 121    Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(4)/LOGICAL(4)).
     122    /tmp/klinveam/spack-stage/spack-stage-pfunit-3.3.3-rasd4ft5hcapocfcxzm5duytvk75klov/spack-src/source/MpiCon
            text.F90:180:25:
     123
     124      180 |       call mpi_allreduce(value, tmp, 1, MPI_INTEGER, MPI_SUM, &
     125          |                         1
     126    ......
     127      325 |       call MPI_Allreduce(q, anyQ, 1, MPI_LOGICAL, MPI_LOR, &
     128          |                         2
  >> 129    Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(4)/LOGICAL(4)).
     130    /usr/bin/cmake -E cmake_copy_f90_mod /home/klinveam/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-10.1.0/pf
            unit-3.3.3-rasd4ft5hcapocfcxzm5duytvk75klov/include/throwfundamentaltypes_mod.mod source/CMakeFiles/pfunit.
            dir/throwfundamentaltypes_mod.mod.stamp GNU
     131    source/CMakeFiles/pfunit.dir/build.make:903: recipe for target 'source/CMakeFiles/pfunit.dir/MpiContext.F90
            .o' failed
  >> 132    make[2]: *** [source/CMakeFiles/pfunit.dir/MpiContext.F90.o] Error 1
     133    make[2]: *** Waiting for unfinished jobs....
     134    make[2]: Leaving directory '/tmp/klinveam/spack-stage/spack-stage-pfunit-3.3.3-rasd4ft5hcapocfcxzm5duytvk75
            klov/spack-build'
     135    CMakeFiles/Makefile2:183: recipe for target 'source/CMakeFiles/pfunit.dir/all' failed
  >> 136    make[1]: *** [source/CMakeFiles/pfunit.dir/all] Error 2
     137    make[1]: Leaving directory '/tmp/klinveam/spack-stage/spack-stage-pfunit-3.3.3-rasd4ft5hcapocfcxzm5duytvk75
            klov/spack-build'
     138    Makefile:173: recipe for target 'all' failed
  >> 139    make: *** [all] Error 2
```

It looks like gfortran 10 is a bit more finicky about datatypes than 9 was.

--- a/var/spack/repos/builtin/packages/pfunit/package.py
+++ b/var/spack/repos/builtin/packages/pfunit/package.py
@@ -44,6 +44,7 @@ class Pfunit(CMakePackage):
     # See https://github.com/Goddard-Fortran-Ecosystem/pFUnit/pull/179
     conflicts("+shared", when="@4.0.0:")
     conflicts("use_comm_world", when="~mpi")
+    conflicts('+mpi', when='@:3.99.99 %gcc@10.0.0:')
     patch("mpi-test.patch", when="@:3.99.99 +use_comm_world")
 
     def patch(self):
@@ -73,10 +74,17 @@ class Pfunit(CMakePackage):
             '-DMAX_RANK=%s' % spec.variants['max_array_rank'].value]
 
         if spec.satisfies('+mpi'):
-            args.extend(['-DMPI=YES', '-DMPI_USE_MPIEXEC=YES',
+            if spec.satisfies('@4.0.0:'):
+                args.append('-DSKIP_MPI=NO')
+            else:
+                args.append('-DMPI=YES')
+            args.extend(['-DMPI_USE_MPIEXEC=YES',
                          '-DCMAKE_Fortran_COMPILER=%s' % spec['mpi'].mpifc])
         else:
-            args.append('-DMPI=NO')
+            if spec.satisfies('@4.0.0:'):
+                args.append('-DSKIP_MPI=YES')
+            else:
+                args.append('-DMPI=NO')
         return args
 
     def check(self):

--- a/var/spack/repos/builtin/packages/pfunit/package.py
+++ b/var/spack/repos/builtin/packages/pfunit/package.py
@@ -73,18 +73,15 @@ class Pfunit(CMakePackage):
             '-DOPENMP=%s' % ('YES' if '+openmp' in spec else 'NO'),
             '-DMAX_RANK=%s' % spec.variants['max_array_rank'].value]
 
+        if spec.satisfies('@4.0.0:'):
+            args.append('-DSKIP_MPI=%s' % ('YES' if '~mpi' in spec else 'NO'))
+        else:
+            args.append('-DMPI=%s' % ('YES' if '+mpi' in spec else 'NO'))
+
         if spec.satisfies('+mpi'):
-            if spec.satisfies('@4.0.0:'):
-                args.append('-DSKIP_MPI=NO')
-            else:
-                args.append('-DMPI=YES')
             args.extend(['-DMPI_USE_MPIEXEC=YES',
                          '-DCMAKE_Fortran_COMPILER=%s' % spec['mpi'].mpifc])
-        else:
-            if spec.satisfies('@4.0.0:'):
-                args.append('-DSKIP_MPI=YES')
-            else:
-                args.append('-DMPI=NO')
+
         return args
 
     def check(self):

